### ISSUE TITLE
weird vertical reading of PDF

### ISSUE THREAD

**Issue by MathieuDerelle** [yob/pdf-reader#316]
using snippet : 

```
p=PDF::Reader.new(path);
ap p.pages.map { |page| page.text.split("\n") }
```

`ap` being [awesome print](https://github.com/awesome-print/awesome_print)

```
[
    [0] [
        [ 0] "a   D                                                                              D    F 0 2 C C A             F",
        [ 1] "r   é                                  0          0          0          C          é    n 7 ,S R i R            a",
        [ 2] "B   a                                  0          0          0          d          t    e 0  Q I s E            t",
        [ 3] "n   i                                  8          1          1          e          i      L  A N i S",
        [ 4] "a   d                                  8          5          2                     d      E  R E S E            e",
        [ 5] "e   u                                  9          4          7                     e      R E  A u D",
        [ 6] "    p                                  9          3          8                     a      I  A N i L",
        [ 7] "    i                                                                              c      E  C L   V",
        [ 8] "    m                                                                              o      S  H E   R",
        [ 9] "    e                                                                              m           R   I",
        [10] "    n                                                                              m               O",
        [11] "    t                                                                              n               N",
        [12] "                                                                                   d",
        [13] "                                                                                   e",
        [14] "",
        [15] "                                    5 D D D    0 B B C    1 E M G       P",
        [16] "                                    0 i  o I   1 B B  l   0 a R  v      r",
        [17] "                                    P l  l R   g S S  n   m d G  n      d                 F 6 6 C A",
        [18] "                                    t s s      t n I  s   l e I  h      u                 n  7 S R R",
        [19] "                                    l A A        D D        o N  y      t                 e  0 Q N E             1",
        [20] "                                    d b b        t E        e C                              F A N S             ,",
        [21] "1                                   e i rc       x O        t H                              E R E E             R",
        [22] ",                                   s t t        l X          Y                              R E A D           5 e M",
        [23] "8                                   e            d                                           I A N E           0 R r",
        [24] "€                                                S                                           E C L A           , a n",
        [25] "                                                 P                                           S H I T           a m a",
        [26] "                                                 2                                               R U           i u d",
        [27] "                                                 5                                                 A             r",
        [28] "                                                                                                   T",
        [29] "                                                                                                   I",
        [30] "               M                                                          U                        N",
        [31] "               n                       7          3          6         T  n P",
        [32] "               t       F               ,          9          9         T  i i",
        [33] "               n       a               9          9          9         C  i x",
        [34] "               t       s               €          €          €                          0 D D N N D",
        [35] "               o   D   d                                                                0 t t °d d T               S",
        [36] "               t   o   e   S                                                Q           . d d  e e A               A",
        [37] "           D   l   t   l   o           1          1          1              é           0 e e  o f S               S",
        [38] "           n   à   r   r   s                                                            9 a com  t D         T     u",
        [39] "           t   a   m   i   T                                                              u  m a r E         A C   a",
        [40] "           V   y   i   o   t           2          2          2                            a  a d : F         n S r i",
        [41] "           A   r   e   n               ,          ,          ,              T             o  d e 0 C         r 3 t a",
        [42] "                           l           0          0          0              A             e  e 3 9 T       F c 4 3 d",
        [43] "                                       %          %          %       c                    t  0 7 2 R       R m 6 8 2",
        [44] "                                                                     o      M             '  . 3 1 A       0 m 7 7 1",
        [45] "               1           1           1          3          6       m (    o             p  6 5 6 O       4 n 1 1 0",
        [46] "           8   1   3   5   1           9          0          0       r m    n             d  0 2 0 N       6 u 9 9 4",
        [47] "           5   2   6   9   2           ,          7          9       i i  T a             i  1 5           4 a P 5 9",
        [48] "           5   8   4   5   8           €          €          €       e e  C n             n  9             6 e a 0 4",
        [49] "           €   €   €   €   €                                         )      t             :                9 : s 5 €"
    ],
    [1] [
        [ 0] "     p N",
        [ 1] "M    o  o",
        [ 2] "a    v  t",
        [ 3] "i    z  S",
        [ 4] "n    a  e",
        [ 5] "u    s  i",
        [ 6] "d    i",
        [ 7] "o       C",
        [ 8] "u    s  l",
        [ 9] "r       t",
        [10] "m    l  e",
        [11] "e    n  s",
        [12] "r    s  à",
        [13] "i       v",
        [14] "o    c  t",
        [15] "u    d",
        [16] "v    i  c",
        [17] "t    n  u",
        [18] "     s  t",
        [19] "c    é  c",
        [20] "n    é n",
        [21] "f    a n",
        [22] "c    e ce",
        [23] "e    d zt",
        [24] "t    e  -",
        [25] "e    e u",
        [26] "p    n s",
        [27] "r    e  s",
        [28] "q    A r",
        [29] "u    u  l",
        [30] "v    u  t",
        [31] "o    n  e",
        [32] "e    e  w",
        [33] "c    o  w",
        [34] "m    m  m",
        [35] "m    t  a",
        [36] "a    n  o",
        [37] "d    '  n",
        [38] "e    t ua",
        [39] "o    a d",
        [40] "s    o  ,",
        [41] "a    d  u",
        [42] "p    é  b",
        [43] "o    .  q",
        [44] "t       e",
        [45] "a       N",
        [46] "n       o",
        [47] "i       s",
        [48] "r       C",
        [49] "s       n",
        [50] "a       a",
        [51] "s       t",
        [52] "c       e",
        [53] "i       V",
        [54] "n       o",
        [55] ".       s"
    ]
]
```

I've sent you the original PDF via mail.

edit : I've tested against current master 9dabde159011b031fc34c078775fd001b87379d2

**Comment by yob** [comment]
huh, that definitely doesn't look right!

I'll try to find some time to look into it. Hopefully the issue is something obvious.

**Comment by yob** [comment]
The sample PDF you provided is an interesting one!

It uses matrix transformations in the page content to render the content at 90 degrees, and then something has modified the page to have an overall 270 degree rotation. The end result is it looks normal to the eye, but there's lots of unnecessary matrix maths involved.

There's definitely a bug in pdf-readers text extraction. I suspect we're not including the 270 page rotation in the maths.

I don't have a fix yet, but I have created a failing spec using a manually crafted file that behaves the same way.

**Comment by yob** [comment]
Gah. #317 helps in some cases like this, but it's not a complete fix for the PDF you sent me. Sorry!

**Comment by yob** [comment]
This is very late, but the positioning issues in the sample file you sent are addressed by the fixes I'm working on for #397. Hopefully I'll land them on main in the next few days and we can resolve this issue at the same time.
